modvendor/modvendor_test.go

```diff
@@ -15,18 +15,823 @@
+			wantIgnored: []wantIgnoredVendor{
+				{
+					RawSource:     "git::file://{{.}}/module-that-does-not-exists?ref=main",
+					ReasonPattern: "failed to vendor",
+				},
+			},
+		},
+		{
+			name: "module with 1 remote dependency",
+			layout: []string{
+				"g:module-test",
+				"g:another-module",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: Module(
+						Labels("test"),
+						Str("source", "git::file://{{.}}/another-module?ref=main"),
+					),
+				},
+			},
+			wantFiles: map[vendorPathSpec]fmt.Stringer{
+				"git::file://{{.}}/module-test?ref=main#main.tf": Module(
+					Labels("test"),
+					Str("source", "{{index . 1}}"),
+				),
+			},
+			wantVendored: []string{
+				"git::file://{{.}}/module-test?ref=main",
+				"git::file://{{.}}/another-module?ref=main",
+			},
+		},
+		{
+			name: "module with 1 remote dependency that contains bogus module.source",
+			layout: []string{
+				"g:module-test",
+				"g:another-module",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: Module(
+						Labels("test"),
+						Str("source", "git::file://{{.}}/another-module?ref=main"),
+					),
+				},
+				{
+					repo: "another-module",
+					path: "another-module/main.tf",
+					data: Module(
+						Labels("test"),
+						Str("source", "git::"),
+					),
+				},
+			},
+			wantFiles: map[vendorPathSpec]fmt.Stringer{
+				"git::file://{{.}}/module-test?ref=main#main.tf": Module(
+					Labels("test"),
+					Str("source", "{{index . 1}}"),
+				),
+			},
+			wantVendored: []string{
+				"git::file://{{.}}/module-test?ref=main",
+				"git::file://{{.}}/another-module?ref=main",
+			},
+			wantIgnored: []wantIgnoredVendor{
+				{
+					RawSource:     "git::",
+					ReasonPattern: "reference must be non-empty",
+				},
+			},
+		},
+		{
+			name: "module with 1 remote dependency referenced multiple times in same file",
+			layout: []string{
+				"g:module-test",
+				"g:another-module",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: Doc(
+						Module(
+							Labels("test"),
+							Str("source", "git::file://{{.}}/another-module?ref=main"),
+						),
+						Module(
+							Labels("test"),
+							Str("source", "git::file://{{.}}/another-module?ref=main"),
+						),
+						Module(
+							Labels("test"),
+							Str("source", "git::file://{{.}}/another-module?ref=main"),
+						),
+						Module(
+							Labels("test"),
+							Str("source", "git::file://{{.}}/another-module?ref=main"),
+						),
+					),
+				},
+			},
+			wantFiles: map[vendorPathSpec]fmt.Stringer{
+				"git::file://{{.}}/module-test?ref=main#main.tf": Doc(
+					Module(
+						Labels("test"),
+						Str("source", "{{index . 1}}"),
```

Comment:
Hmm if there was a way for these references to be by name, like the name of the module you depend and then it expands do the source, it would improve things considerably, the positional access is quite hard to understand for a newcomer (or someone as limited as me :laughing: ...I think I get it now, and the problem is tricky, we can assess this later).

Reply:
Ahh...would be already an improvement. We can address in the future :+1: